Addition of Dirichlet, Ewens Distributions

czgdp1807 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
[1] [GSoC Proposal](https://docs.google.com/document/d/1oIeaROiJyglpbris7X1uZPRE5ZeO0pD1ygCFhBIeATI/edit#heading=h.wb8kk2gus2ng) by @czgdp1807 
[2] [GSoC Proposal](https://github.com/ritesh99rakesh/gsocProposal/blob/master/SymPy.md) by @ritesh99rakesh 

#### Brief description of what is fixed or changed
MultivariateBetaDistribution has been added to sympy.stats.joint_rv_types. More distributions will be added after reviewing the previous one added. 

#### Other comments
I want to use `Beta` in `sympy.stats.crv_types` for making random symbols following `MultivariateBetaDistribution` as is `Normal` being used for `MultivariateNormalDistribution`. However, the reason restricting me for using `Beta` is the difference in the way parameters are passed to the `Beta` function and `MultivariateBeta` function. The former one accepts `alpha`, `beta` as arguments and the latter one accepts, a single vector or pythonically, list, `alpha`. One solution in my mind is to use, `Beta(name, *alpha)`, for both univariate and multivariate Beta distributions. The length of and type of `*alpha` can be used as the deciding factor for returning either return `multivariate_rv` or `rv`. 

Edit - As said in [this comment](https://github.com/sympy/sympy/pull/16576#issuecomment-491218778) below, I think using `Beta` for `MultivariateBeta` or `Dirichlet` distribution is not a good idea because, most people use `Dirichlet` to denote `MultivariateBetaDistribution`, so mixing up the two distributions is not valid. However, both, `MultivariateBeta` and `Dirichlet` can be used to denote `MultivariateBetaDistributions` . 

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* stats
  * multivariate distributions added to sympy.stats.joint_rv_types
<!-- END RELEASE NOTES -->

czgdp1807 committed:
Added MultivariateBetaDistribution

MultivariateBetaDistribution has been added
to sympy.stats.joint_rv_types. Tests and
Documentation will be added in the next commit.

czgdp1807 committed:
Added documentation and tests for MultivariateBeta

czgdp1807 committed:
removed some unwanted changes

Sc0rpi0n101 commented:
The build notes say you have to add a test for the new class in test_args.py

czgdp1807 committed:
Added tests in test_args

czgdp1807 commented:
@jksuom @Upabjojr Please review this.
It would be great if we can discuss on the other comments section in the [description](https://github.com/sympy/sympy/pull/16576#issue-268051524).

sidhantnagpal commented on sympy/stats/joint_rv_types.py:
```diff
@@ -253,3 +253,71 @@ def NormalGamma(syms, mu, lamda, alpha, beta):
+    def pdf(self, *syms):
+        alpha = self.alpha
+        B = Mul(*[gamma(a_k) for a_k in alpha])/gamma(Add(*[a_k for a_k in alpha]))
+        return Mul(*[sym**(a_k) for a_k, sym in zip(alpha, syms)])/B
```
The exponent should be `a_k - 1`?

sidhantnagpal commented on sympy/stats/joint_rv_types.py:
```diff
@@ -253,3 +253,71 @@ def NormalGamma(syms, mu, lamda, alpha, beta):
+
+    def pdf(self, *syms):
+        alpha = self.alpha
+        B = Mul(*[gamma(a_k) for a_k in alpha])/gamma(Add(*[a_k for a_k in alpha]))
```
This could possibly be simplified.
```suggestion
        B = Mul(*list(map(gamma, alpha)))/gamma(Add(*alpha))
```

sidhantnagpal commented on sympy/stats/joint_rv_types.py:
```diff
@@ -253,3 +253,71 @@ def NormalGamma(syms, mu, lamda, alpha, beta):
+    >>> from sympy.stats.joint_rv_types import MultivariateBeta
+    >>> from sympy import Symbol
+    >>> a1 = Symbol('a1', positive=True, real=True)
+    >>> a2 = Symbol('a2', positive=True, real=True)
```
`real=True` can be dropped while defining `a1` and `a2` as they are already implied by `positive=True`.

sidhantnagpal reviewed (commented):
Seems like a good start. I would have expected `MultivariateBeta('B', a_1, a_2, ...)` to work, but I would wait to see what someone with more experience in `stats` thinks about the API.

czgdp1807 committed:
incorporated reviews

czgdp1807 committed:
made changes to the api

czgdp1807 commented:
@sidhantnagpal I have made the changes. 
About API:
I also agree that `MultivariateBeta('B', a_1, a_2, ...)` should work. Therefore, I have added the provision of both passing the parameters in the form of list as well as in the format that you suggested. I think @Upabjojr will also agree to such an API. If not, then please let me know which API should I go for?

czgdp1807 commented:
@Abdullahjavednesar Please add the `GSoC` label on this PR.

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -253,3 +253,74 @@ def NormalGamma(syms, mu, lamda, alpha, beta):
+
+    def check(self, alpha):
+        _value_check(len(alpha) >= 2, "At least two categories should be passed.")
+        _value_check(all([a_k > 0 for a_k in alpha]), "Each concentration parameter"
```
this will raise an exception if `a_k` is symbolic. `a_k > 0` creates a `GreaterThan` object which cannot be cast to a boolean.

I usually do  `(a_k > 0) == True`.

jksuom replied:
I think that `_value_check` should be able to handle also Relational type conditions. It has `if condition == False`.

Upabjojr replied:
Maybe ` (a_k > 0) != False` is the correct condition. You would probably allow generic symbols.

Upabjojr replied:
```python
for a_k in alpha:
    _value_check(a_k > 0, "...")
```

czgdp1807 replied:
@jksuom @Upabjojr I think in `_value_check` the error message should be raised when `condition == None` apart from `condition == False`. As in the following session, `theta.is_positive` is `None` but it will get through checks which require theta to be positive and not negative. 
```
Python 3.6.7 (default, Oct 22 2018, 11:32:17) 
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> theta = symbols('theta', real = True)
>>> theta.is_positive
>>> theta.is_negative
>>> theta.is_real
True
``` 
PS - I haven't tried the change but may be it can require change at lots of places but I believe that efforts required hold little value when concepts need to be corrected. Though that is just my point of view.

smichr replied:
If `_value_check` is made to return True if it doesn't raise then this can be written `if all(_value_check(a_k < 0, msg) for a_k in alpha):`

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -253,3 +253,74 @@ def NormalGamma(syms, mu, lamda, alpha, beta):
+
+    def pdf(self, *syms):
+        alpha = self.alpha
+        B = Mul(*list(map(gamma, alpha)))/gamma(Add(*alpha))
```
use `Mul.fromiter(iterator)` to generate a multiplication from a list or iterator. Same for `Add` (i.e. `Add.fromiter(alpha)`).

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -253,3 +253,74 @@ def NormalGamma(syms, mu, lamda, alpha, beta):
+    def pdf(self, *syms):
+        alpha = self.alpha
+        B = Mul(*list(map(gamma, alpha)))/gamma(Add(*alpha))
+        return Mul(*[sym**(a_k - 1) for a_k, sym in zip(alpha, syms)])/B
```
`Mul.fromiter`

Upabjojr commented on sympy/stats/tests/test_joint_rv.py:
```diff
@@ -53,6 +53,18 @@ def test_NormalGamma():
+    raises(ValueError, lambda: MultivariateBeta('b1', [1, -2]))
+    raises(ValueError, lambda: MultivariateBeta('b2', [0, 2]))
+    raises(ValueError, lambda: MultivariateBeta('b3', [0, 0]))
+    raises(ValueError, lambda: MultivariateBeta('b4', [-1, -2]))
```
tests for symbolic alphas are probably needed.

Upabjojr commented:
This distribution is usually referred to as _Dirichlet distribution_. Is there a reason why you prefer _multivariate beta_?

Could you please add some references in the documentations (e.g. a link to Wikipedia or MathWorld)?

czgdp1807 commented:
@Upabjojr Thank you for the reviews. From yours and Sidhant's reviews, I conclude  that the current API being used is fine.
[This line](https://github.com/sympy/sympy/pull/16576/files#diff-ce1eb95f95a9cf137ee09ced0921b979R326) allows both `Dirichlet` and `MultivariateBeta`. [Wikipedia of Dirichlet Distribution](https://en.wikipedia.org/wiki/Dirichlet_distribution) suggests that alternative name being used is MultivariateBeta, so I thought of using both. I will add more references in the documentation.

czgdp1807 committed:
Added Ewens Distribution

Multivariate Ewens Distribution has been added
to sympy.stats.joint_rv_types.Few changes to limit
format have also been made in marginal_distribution
of discrete joint distribution.

czgdp1807 commented:
@Upabjojr Please review the changes I have made in the latest commit.
Brief Summary of Changes:
1. Added Multivariate Ewens Distribution. Few unused imports are there, I will remove them when I will add the next multivariate distribution.
2. I have changed the format of limits being passed to summation while computing `marginal_distributions` of discrete joint distributions.

sidhantnagpal commented on sympy/stats/tests/test_joint_rv.py:
```diff
@@ -53,6 +53,32 @@ def test_NormalGamma():
+    from sympy.stats.joint_rv_types import MultivariateBeta
+    from sympy import gamma
+    mb = MultivariateBeta('B', [1, 2])
+    mb_c = MultivariateBeta('C', 1, 2)
```
It would nice to have test(s) with symbolic concentration parameters, currently such tests are only included as doctests.

sidhantnagpal commented on sympy/stats/joint_rv_types.py:
```diff
@@ -253,3 +253,144 @@ def NormalGamma(syms, mu, lamda, alpha, beta):
+
+    def pdf(self, *syms):
+        alpha = self.alpha
+        B = Mul.fromiter(list(map(gamma, alpha)))/gamma(Add(*alpha))
```
Can casting into `list` be dropped as `fromiter` is being used?

sidhantnagpal commented on sympy/stats/tests/test_joint_rv.py:
```diff
@@ -53,6 +53,32 @@ def test_NormalGamma():
+    from sympy import gamma
+    mb = MultivariateBeta('B', [1, 2])
+    mb_c = MultivariateBeta('C', 1, 2)
+    a1_f, a2_f = symbols('a1, a2', positive = False)
```
No spaces around `=` when specifying a keyword argument (here and in other places).

czgdp1807 committed:
incorporated reviews

czgdp1807 committed:
Merge branch 'master' into multivar_dists

czgdp1807 committed:
removed unused import

czgdp1807 committed:
added comments

czgdp1807 commented:
@sidhantnagpal Changes have been made according to the reviews. If there is anything else to change then please comment.
Rest of the distributions will be made in two more PRs.

sidhantnagpal commented on sympy/stats/tests/test_joint_rv.py:
```diff
@@ -53,6 +53,32 @@ def test_NormalGamma():
+    n, theta = symbols('n theta', positive = True)
+    theta_f = symbols('t_f', negative = True)
+    ed = MultivariateEwens('E', 3, 1)
+    assert density(ed)(0, 1, 2) == Rational(1, 36)
```
This doesn't seem right. Note that, `\sum_{k} k a_{k} = n`. This should be taken care of in `pdf`.

czgdp1807 replied:
Yes, I am thinking of using `Piecewise`, will update if it works well.

sidhantnagpal commented on sympy/stats/tests/test_joint_rv.py:
```diff
@@ -53,6 +53,32 @@ def test_NormalGamma():
+    a1_f, a2_f = symbols('a1, a2', positive=False)
+    mb = MultivariateBeta('B', [a1, a2])
+    mb_c = MultivariateBeta('C', a1, a2)
+    assert simplify(density(mb)(1, 2) -
```
Can `simplify` be avoided here (and similarly below)?

czgdp1807 replied:
I used `simplify` as it was in the FAQ, [Why does SymPy say that two equal expressions are unequal?](https://github.com/sympy/sympy/wiki/Faq#why-does-sympy-say-that-two-equal-expressions-are-unequal).

sidhantnagpal replied:
It can be slow as it tries numerous simplifications. It is better if the output can be compared directly without having to using `simplify`.

czgdp1807 replied:
Since, this is just some sympy code at work, I have removed `simplify`. If there will be any problem, we can handle it at that time, for now the performance of the code matters.

sidhantnagpal commented on sympy/stats/joint_rv_types.py:
```diff
@@ -253,3 +253,147 @@ def NormalGamma(syms, mu, lamda, alpha, beta):
+
+    def pdf(self, *syms):
+        n, theta = self.n, self.theta
+        term_1 = factorial(n)/Mul.fromiter([theta + i for i in range(n)])
```
Can `rf(theta, n)` (rising factorial) be used in the denominator?

czgdp1807 replied:
Thanks for the suggestion. I have updated the code locally, will update the PR soon.

sidhantnagpal commented on sympy/stats/tests/test_joint_rv.py:
```diff
@@ -53,6 +53,32 @@ def test_NormalGamma():
+def test_MultivariateEwens():
+    from sympy.stats.joint_rv_types import MultivariateEwens
+    n, theta = symbols('n theta', positive = True)
+    theta_f = symbols('t_f', negative = True)
```
`negative=True` and similar.

sidhantnagpal replied:
Test(s) for symbolic `n` and `theta` with valid assumptions should also be included.

czgdp1807 replied:
@sidhantnagpal  Tests can be added for symbolic `theta` but for `n` it is not possible. The reason for my claim is based on the fact that we can determine `n` from the the length of `syms`. So, one way or the other `n` is an `Integer` and not a `Symbol` and I thought of accepting a fixed `n` from the user rather than figuring out what `n` is from the length of  `syms`.

czgdp1807 committed:
pdf of ewens dist changed

czgdp1807 commented:
@sylee957 Please add the `GSoC` label here too. Thanks.

Upabjojr commented:
Master needs to be merged.

czgdp1807 committed:
Merge branch 'master' into multivar_dists

czgdp1807 committed:
resolved conflicts

czgdp1807 commented:
@sidhantnagpal @Upabjojr Are there any more changes to make apart from the un-evaluated lists. If not, then I think for the time being we can go for only, `Integer` dimensions and once something like `List` is developed we can make separate PR for improving on that part.
